### ISSUE TITLE
Increased backend errors

### ISSUE THREAD

**Issue by GoogleCodeExporter** [runt18/google-bigquery#339]
```
Since Thursday mid day, central time, we have been experiencing larger number 
of backend errors than usual.

response: {"code":503,"errors":[{"domain":"global","message":"Error encountered 
during execution. Retrying may solve the 
problem.","reason":"backendError"}],"message":"Error encountered during 
execution. Retrying may solve the problem."}

I attached a graph that represents an increasing log of data we send to 
bigquery (write pointer), and the "read" pointer shows the process that is 
inserting into Bigquery using bigquery insertAll API.  Normally the insertion 
process is slightly behind the queuing process.  But as you can see in the 
graph, starting mid thursday backend errors start causing the flow to lag, and 
now it is actually getting worse and diverging more than previously.
```

Original issue reported on code.google.com by `lukefore...@gmail.com` on 2 Oct 2015 at 10:32

Attachments:
- [Screenshot 2015-10-02 17.06.37.png](https://storage.googleapis.com/google-code-attachments/google-bigquery/issue-339/comment-0/Screenshot 2015-10-02 17.06.37.png)

**Comment by GoogleCodeExporter** [comment]
```
We attempted to restart our process and now things have gotten MUCH worse.  We 
are basically down in production
```

Original comment by `lukefore...@gmail.com` on 2 Oct 2015 at 10:39

**Comment by GoogleCodeExporter** [comment]
```
Things are starting to pick up a little...
```

Original comment by `lukefore...@gmail.com` on 2 Oct 2015 at 10:52

**Comment by GoogleCodeExporter** [comment]
```
Here is a zoomed in graph of today
```

Original comment by `lukefore...@gmail.com` on 2 Oct 2015 at 11:06

Attachments:
- [Screenshot 2015-10-02 18.06.05.png](https://storage.googleapis.com/google-code-attachments/google-bigquery/issue-339/comment-3/Screenshot 2015-10-02 18.06.05.png)

**Comment by GoogleCodeExporter** [comment]
```
We flatlined around 7:10PM central
```

Original comment by `lukefore...@gmail.com` on 3 Oct 2015 at 3:53

Attachments:
- [Screenshot 2015-10-02 22.52.32.png](https://storage.googleapis.com/google-code-attachments/google-bigquery/issue-339/comment-4/Screenshot 2015-10-02 22.52.32.png)

**Comment by GoogleCodeExporter** [comment]
```
Can you please provide the name of the project?
```

Original comment by `epa...@google.com` on 3 Oct 2015 at 4:32

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
pioneering-tome-782.

as of a few minutes ago we achieved over 1 million record acks per 10
minutes which is a good sign
```

Original comment by `lukefore...@gmail.com` on 3 Oct 2015 at 4:36

**Comment by GoogleCodeExporter** [comment]
```
Now we are achieving over double the rates previously, up to 1,506,440 records 
in 10 minutes.  Something has been un-wedged.
```

Original comment by `lukefore...@gmail.com` on 3 Oct 2015 at 4:39

**Comment by GoogleCodeExporter** [comment]
```
We noticed an increase of 404s for your project earlier this evening (possibly 
around the time you restarted your process). That seems to have gone down now, 
but we haven't seen any significant increase of 503 for your project.

Please let us know if this  continues or happens again.

Thank you.
```

Original comment by `epa...@google.com` on 3 Oct 2015 at 5:26

**Comment by GoogleCodeExporter** [comment]
```
Some additional information that will perhaps be useful:

From our side, we did see an error spike at the request and row level at ~1pm 
central. Otherwise, our logging for your project indicates a fairly low error 
rate for the rest of the day. Latencies were also well within expected 
boundaries. It does look like there was a dramatic drop in requests received 
from your system around 7pm central, until about 10pm (the error request spike 
appears to be for many NOT_FOUND errors, usually related to missing tables). 

In any case, our the row rates seems to scale fairly closely with the valid 
request rates your system is sending to us. Also, our system as a whole did not 
see an equivalent drop/degradation requests/rows accepted during those time 
periods (though there was a portion of the system that did see the ~1pm error 
spike), and is generally handling orders of magnitude more requests/rows. Is it 
possible there's a bottleneck on the request generation side somewhere?

(day_hour is in central time... Formatting may make some values appear joined 
with a neighbor)
Row day_date    day_hour    requests    error_requests  total_rows  error_rows
36  2015-10-02  4   488446  48  9139740 78
37  2015-10-02  5   473433  34  9533953 10
38  2015-10-02  6   594524  24  12275933    1
39  2015-10-02  7   451035  30  9260483 6
40  2015-10-02  8   434848  16  9137776 null
41  2015-10-02  9   823615  51  15644375    6
42  2015-10-02  10  407657  57  7530771 9
43  2015-10-02  11  420721  32  7674634 4
44  2015-10-02  12  402379  28  7124921 29
45  2015-10-02  13  365006  172 6369541 2044
46  2015-10-02  14  438299  80  7288949 32
47  2015-10-02  15  440971  47  7143838 6
48  2015-10-02  16  401475  46  6814285 29
49  2015-10-02  17  492750  20  8150190 26
50  2015-10-02  18  102448  2904    1988946 14
51  2015-10-02  19  50802   4936    1140034 3
52  2015-10-02  20  48973   5054    1118837 null
53  2015-10-02  21  160585  3180    2692334 null
54  2015-10-02  22  951123  34  13980825    21
55  2015-10-02  23  965239  30  14109833    13
```

Original comment by `seanc...@google.com` on 3 Oct 2015 at 9:18

**Comment by GoogleCodeExporter** [comment]
```
We did have a problem with attempting to insert into missing (expired) tables.  
That was actually a separate issue.  Once resolved, we were no longer 
cratering, but we are still unable to push past a certain rate of around 750K 
per 10 minutes, even after doubling the number of connections from 18 to 36, 
for each dataset.
```

Original comment by `lukefore...@gmail.com` on 3 Oct 2015 at 5:11

**Comment by GoogleCodeExporter** [comment]
```
Perhaps if you could describe your architecture a bit, it would help evaluate 
things. Does connections equal the number of threads on a single server, the 
number of distinct servers performing inserts, etc? From what I can tell you're 
nowhere near our rate-limits (and we don't see any rate-limiting failures).

For comparison, we have active projects streaming at the 10s to 100s of 
millions of rows per 10 minute interval.
```

Original comment by `seanc...@google.com` on 3 Oct 2015 at 5:58

**Comment by GoogleCodeExporter** [comment]
```
It is storm, one connection per thread.  I increased threads but not servers 
perhaps I should try that. The request is a batch request of insertAll 
requests, each to a different table.  The payload is very large sometimes, we 
hit your record limits regularly and eject records that exceed the limit
```

Original comment by `lukefore...@gmail.com` on 3 Oct 2015 at 6:52

**Comment by GoogleCodeExporter** [comment]
```
Here is an update:  I have doubled the number of workers in the topology, from  
6 to 12, and the issue remains.  Prior to doubling workers, I had doubled 
number of connections from 18 to 36 for each dataset (we have 4 datasets).
```

Original comment by `lukefore...@gmail.com` on 5 Oct 2015 at 3:50

Attachments:
- [Screenshot 2015-10-02 22.52.32.png](https://storage.googleapis.com/google-code-attachments/google-bigquery/issue-339/comment-14/Screenshot 2015-10-02 22.52.32.png)

**Comment by GoogleCodeExporter** [comment]
```
Profiling one of the JVMs that is streaming to Bigquery API, it looks like the 
primary hotspot in the process is waiting for HttpResponse, would that indicate 
an issue on your end?
```

Original comment by `lukefore...@gmail.com` on 5 Oct 2015 at 4:19

Attachments:
- [Screenshot 2015-10-05 11.15.54.png](https://storage.googleapis.com/google-code-attachments/google-bigquery/issue-339/comment-15/Screenshot 2015-10-05 11.15.54.png)
- [Screenshot 2015-10-05 11.19.23.png](https://storage.googleapis.com/google-code-attachments/google-bigquery/issue-339/comment-15/Screenshot 2015-10-05 11.19.23.png)

**Comment by GoogleCodeExporter** [comment]
```
Thanks for the updates Luke.

With respect to your first post -- That's unfortunate! It could be an issue on 
either side, though (as mentioned before) we really don't believe it's a 
generalized problem with our system, as we currently have single tables 
receiving higher rates of ingestion that your entire project.

With respect to your second post, it's not clear either way. If your process 
doesn't do much besides translated and issue http requests, then it could be 
completely normal (since network requests will be of a time scale much larger 
than any local computation).

I do have a follow up questions/comments based on your earlier response:

When you say a batch of insertAll requests, I assume you mean the google apis 
batch request mechanism? We (BigQuery) don't monitor at that level (it's part 
of a higher-level infrastructure), so it's possible that that system is adding 
latency. Is it possible for you to unwind that batching, and simply issue a 
request per table? You can group multiple rows for the same destination in a 
single request, but issuing each request directly may do a couple things:

1) More granular work, and thus long tail latencies in inserts to one 
destination don't stall all other requests.
2) Reduce the incidence of "request too large". I believe the higher level 
system has a limit on HTTP request size. It's possible that using the batch 
request mechanism hits that limit much sooner.

Of course in doing this, you may want to increase your thread count (should be 
ok, as those threads should mostly be blocked waiting on IO).


Finally, do you have any stats on your request/response timing and success 
ratios etc?

```

Original comment by `seanc...@google.com` on 5 Oct 2015 at 4:34

**Comment by GoogleCodeExporter** [comment]
```
Yes we are using the google apis batch request mechanism.  I am guessing 
something might have changed at the higher level infrastructure, since we have 
made no changes to cause an issue.  By the way, we do detect "request too 
large" issue before sending a request; we make sure our batch size respects 
this limit.

I might try unwinding the  batches
```

Original comment by `lukefore...@gmail.com` on 5 Oct 2015 at 4:46

**Comment by GoogleCodeExporter** [comment]
```
We don't have a great report right now for request/response/success timings, I 
will have to get back to you on that.
```

Original comment by `lukefore...@gmail.com` on 5 Oct 2015 at 4:48

**Comment by GoogleCodeExporter** [comment]
Original comment by `thomasp...@google.com` on 24 Nov 2015 at 7:21
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Closing for now, as there hasn't been any activity in ~2 months. Please reopen 
if you get more details etc.
```

Original comment by `seanc...@google.com` on 15 Dec 2015 at 10:19
- Changed state: **WontFix**
